candles: Respect the cursor location when copying.

JoeGruffins opened the pull request:
closes #2168

JoeGruffins commented on dex/candles/candles.go:
```diff
@@ -101,10 +101,27 @@ func (c *Cache) WireCandles(count int) *msgjson.WireCandles {
+	candles := make([]msgjson.Candle, sz)
+	switch {
+	case sz == 0:
+	case sz == c.cap: // circular mode
+		oldIdx := c.cursor + 1
+		copy(candles, c.Candles[oldIdx:])
+		copy(candles[sz-oldIdx:], c.Candles)
```
Is this wrong if it is called just after the size becomes c.cap but before it starts to reuse index 0?

JoeGruffins replied:
If len is 1000 then cursor is at 999... so ok I think.

chappjc replied:
Looks to work right.  Could make some unit tests to explore the edges to be sure about it.  Also we could wrap like done in `Add` with `oldIdx := (c.cursor + 1) % c.cap` so `oldIdx` really is the index of the oldest candle as the name implies?

norwnd reviewed (approved):
Running without gaps for me so far, and my fiddling with the code nearby supports that this is the root cause, nice find!

norwnd commented:
I left `6a93d24z` commit running with **mainnet** overnight (but closed laptop lid before went to bed), now I opened the lid and the gap was there for ~2-3 mins maybe (but resolved now on it's own, without dexc restart or anything):

[candle-gaps.webm](https://user-images.githubusercontent.com/112318969/222316367-91ec6b4a-84b6-460c-9c64-ff40f4dda42d.webm)

full log: 
[candles-gap-lid-closed.txt](https://github.com/decred/dcrdex/files/10867042/candles-gap-lid-closed.txt)

chappjc commented:
> I left `6a93d24z` commit running with **mainnet** overnight (but closed laptop lid before went to bed), now I opened the lid and the gap was there for ~2-3 mins maybe (but resolved now on it's own, without dexc restart or anything):
> 
> full log: [candles-gap-lid-closed.txt](https://github.com/decred/dcrdex/files/10867053/candles-gap-lid-closed.txt)

I'll take that as a win.  😛

JoeGruffins commented:
Resolving itself is an improvement. Not great that there was a gap at all. I guess the issue is not completely solved? Do those logs end after it resolved itself? Maybe it was actually just a few minutes waiting for candles to come it or something?

JoeGruffins commented:
Why are your web and core times so off? They line up in my logs.

```
2023/03/02 03:23:43 "GET http://192.168.1.67:5757/api/user HTTP/1.1" from xxx - 200 50241B in 5.136757ms
2023/03/02 03:23:43 "GET http://192.168.1.67:5757/img/coins/dcr.png HTTP/1.1" from xxx - 200 2909B in 1.229385ms
2023/03/02 03:23:43 "GET http://192.168.1.67:5757/img/coins/ltc.png HTTP/1.1" from xxx - 200 2181B in 667.34µs
2023/03/02 03:23:43 "GET http://192.168.1.67:5757/img/coins/btc.png HTTP/1.1" from xxx - 200 3636B in 864.41µs
2023/03/02 03:23:43 "GET http://192.168.1.67:5757/ws HTTP/1.1" from xxx - 000 0B in 657.914µs
2023-03-02 03:23:43.660 [DBG] WEB[WS]: New websocket client xxx
2023-03-02 03:23:43.660 [TRC] WEB[WS][xxx]: Starting websocket messaging with peer xxx
2023-03-02 03:23:43.664 [TRC] WEB[WS]: message of type 1 received for route loadmarket
2023-03-02 03:23:43.664 [TRC] WEB[WS]: message of type 1 received for route loadcandles
```

norwnd commented:
> Do those logs end after it resolved itself? 

I think logs attached above should include that "auto-resolution", however just to be sure here are the logs continuing further (well after the issue is resolved, and I just terminated dexc):  [candles-gap-lid-closed.txt](https://github.com/decred/dcrdex/files/10867241/candles-gap-lid-closed.txt)

> Not great that there was a gap at all. I guess the issue is not completely solved? Maybe it was actually just a few minutes waiting for candles to come in or something?

(as a user) I'd say it doesn't bother me personally too much, but if there is a way to improve it - sync cache faster or whatnot - I'd appreciate it (and it also could manifest in different ways we aren't currently aware of too).

> Why are your web and core times so off @norwnd ? They line up in my logs.

Might wanna look at why this is too ... it might confuse at times. Looks like WEB logs with "local time zone" while CORE logs with UTC.

chappjc commented:
> Not great that there was a gap at all. I guess the issue is not completely solved?

The computer was suspended.  Definitely resolved

JoeGruffins commented:
> Looks like WEB logs with "local time zone" while CORE logs with UTC.

Do you have a flag set? I think I am on default settings and don't see a difference.

JoeGruffins commented:
Added a test. Removed an unused constant in another file. Added `% c.cap` to the old index calculation just to be safe. https://github.com/decred/dcrdex/compare/93fea5db0e4f64f0f68062381ae04498a0412a10..8ce59c63ce0983564b078ca814488c6d8fa6d3e8
